File: domain/src/main/java/org/openmedstack/eventstore/ConflictDetector.java
package org.openmedstack.eventstore;

import org.openmedstack.Tuple;

import java.util.HashMap;
import java.util.function.Function;

public class ConflictDetector implements IDetectConflicts {
    private final HashMap<Class, HashMap<Class, Function<Tuple<Object, Object>, Boolean>>> _actions = new HashMap<>();

    public <TUncommitted, TCommitted> void register(Function<Tuple<TUncommitted, TCommitted>, Boolean> handler) {
        try {
            var typeVariables = this.getClass().getDeclaredMethod("register").getTypeParameters();
            var uncommitted = Class.forName(typeVariables[0].getTypeName());
            var committed = Class.forName(typeVariables[1].getTypeName());
            if (!_actions.containsKey(uncommitted)) {
                _actions.put(uncommitted, new HashMap<>());
            }
            _actions.get(uncommitted).put(committed, (Tuple<Object, Object> t) -> handler.apply((Tuple<TUncommitted, TCommitted>) t));
        } catch (NoSuchMethodException | ClassNotFoundException n) {
        }
    }

    public Boolean conflictsWith(Object[] uncommittedEvents, Object[] committedEvents) {
        for (var uncommittedEvent : uncommittedEvents) {
            for (var committedEvent : committedEvents) {
                if (conflicts(uncommittedEvent, committedEvent)) {
                    return true;
                }
            }
        }
        return false;
    }

    private Boolean conflicts(Object uncommitted, Object committed) {
        if (!_actions.containsKey(uncommitted.getClass()))
        //.TryGetValue(uncommitted.GetType(), out var dictionary))
        {
            return uncommitted.getClass() == committed.getClass();
        }

        var map = _actions.get(uncommitted.getClass());
        if (!map.containsKey(committed.getClass())) {
            return true;
        }

        return map.get(committed.getClass()).apply(new Tuple<>(uncommitted, committed));
    }
}

File: domain/src/main/java/org/openmedstack/eventstore/ConflictDetector.kt
package org.openmedstack.eventstore

import org.openmedstack.Tuple
import java.util.function.Function

class ConflictDetector : IDetectConflicts {
    private val _actions: HashMap<Class<*>, HashMap<Class<*>, Function<Tuple<Any, Any>, Boolean>>> = HashMap()
    override fun <TUncommitted, TCommitted> register(handler: Function<Tuple<TUncommitted, TCommitted>, Boolean>) {
        try {
            val typeVariables = this.javaClass.getDeclaredMethod("register").typeParameters
            val uncommitted = Class.forName(typeVariables[0].typeName)
            val committed = Class.forName(typeVariables[1].typeName)
            if (!_actions.containsKey(uncommitted)) {
                _actions[uncommitted] = HashMap()
            }
            _actions[uncommitted]!![committed] =
                Function { t: Tuple<Any, Any> -> with(handler) { apply(Tuple(t.a as TUncommitted, t.b as TCommitted)) } }
        } catch (n: NoSuchMethodException) {
        } catch (n: ClassNotFoundException) {
        }
    }

    override fun conflictsWith(uncommittedEvents: Array<Any>, committedEvents: Array<Any>): Boolean {
        for (uncommittedEvent in uncommittedEvents) {
            for (committedEvent in committedEvents) {
                if (conflicts(uncommittedEvent, committedEvent)) {
                    return true
                }
            }
        }
        return false
    }

    private fun conflicts(uncommitted: Any, committed: Any): Boolean {
        if (!_actions.containsKey(uncommitted.javaClass)) //.TryGetValue(uncommitted.GetType(), out var dictionary))
        {
            return uncommitted.javaClass == committed.javaClass
        }
        val map = _actions[uncommitted.javaClass]
        return if (!map!!.containsKey(committed.javaClass)) {
            true
        } else map!![committed.javaClass]!!.apply(Tuple(uncommitted, committed))
    }
}